### ISSUE TITLE
[Enhancement]: information_schema queries should be avoided

### ISSUE THREAD

**Issue by prettyboymp** [woothemes/woocommerce#37276]
### Describe the solution you'd like

Querying against information_schema breaks when using advanced database interfaces like HyperDB.

Many of the advanced database interfaces rely on the table name returned from `wpdb::get_table_from_query()` in order to determine what data server to send the query to.  However, WordPress core's `wpdb::get_table_from_query()` doesn't currently support parsing `information_schema` queries.  

Most places where information_schema is being queried can be avoided by using alternative approaches.

There is a ticket for core to improve it, but it hasn't had much traction and doesn't appear to have a clear path forward for dealing with these.  https://core.trac.wordpress.org/ticket/50789

**Comment by psealock** [comment]
Is the enhancement WooCommerce specific? If not, it would probably be best to advance the ticket in Core somehow.

**Comment by rrennick** [comment]
> Is the enhancement WooCommerce specific?

@psealock Yes, there are 4 update routines that query information_schema.

**Comment by prettyboymp** [comment]
> it would probably be best to advance the ticket in Core somehow.

Yes and no.  Part of the reason the WordPress core ticket stalled was because there isn't a clear path forward.  The  `wpdb::get_table_from_query()` method was originally copied from HyperDB to `wpdb`, but for a different purpose, to determine the table so it could determine what charset. In order to add support for `information_schema` queries, each of the DB interface plugins probably need to (re-)add their own method to determine the table from the query. 
 
If the ability to determine the target table was added to these HyperDB, there would still be an issue with the current `information_schema` queries being used.  As an example, see `wc_update_700_remove_download_log_fk()`

https://github.com/woocommerce/woocommerce/blob/9ea8b630eed18028ed826b40540a9950321f28b4/plugins/woocommerce/includes/wc-update-functions.php#L2451-L2467

There are a couple concerns with this.  First, `$wpdb->dbname` is protected in the current version of WordPress core, so we can't set the `CONSTRAINT_SCHEMA`.  However, if it were public, the query still wouldn't be compatible with one of the DB interfaces because the DB interfaces can't determine the database until they know the table, which it parses from the query.

As it is, with the `CONSTRAINT_SCHEMA` set in the query, the query itself has an `EXPLAIN` description of `Using where; Open_full_table; Scanned 1 database`.   If we removed the `CONSTRAINT_SCHEMA`, the performance would be even worse by changing to `Using where; Open_full_table; Scanned all databases`.

It is best to avoid the `information_schema` queries when possible.
